Add dfba package

Midnighter opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/dfba```) and found some lint.

Here's what I've got...


For **recipes/dfba**:

 * The recipe must have some tests.
 * Non noarch packages should have python requirement without any version constraints.
 * Non noarch packages should have python requirement without any version constraints.

For **recipes/dfba**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

Documentation on acceptable licenses can be found [here]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#spdx-identifiers-and-expressions ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/dfba```) and found some lint.

Here's what I've got...


For **recipes/dfba**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/dfba```) and found it was in an excellent condition.

davidtourigny commented:
Can confirm I am a maintainer of this recipe

Midnighter commented:
@davidtourigny this is starting to look better 😃 Success on Linux and Mac. So we can either skip Windows for now or we need to implement two changes:

1. The NMAKE build system on Win is slightly different. I can edit the build steps.
2. We should add a test for on-the-fly compilation on Windows and we would need the dll shenanigans. Are you up for implementing that?

davidtourigny commented:
> @davidtourigny this is starting to look better 😃 Success on Linux and Mac. So we can either skip Windows for now or we need to implement two changes:
> 
>     1. The NMAKE build system on Win is slightly different. I can edit the build steps.
> 
>     2. We should add a test for on-the-fly compilation on Windows and we would need the dll shenanigans. Are you up for implementing that?

Great! I can certainly begin the implementation-- I think it won't be too difficult but of course something unexpected could pop up. Does it make sense skipping the windows build in the meantime to avoid excessive delays?

Midnighter commented:
@conda-forge/help-python-c please review this addition.

dopplershift commented on recipes/dfba/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+  build:
+    - cmake
+    - {{ compiler('cxx') }}
+    - python
```
Should be enough to have `python` under `host`, so can remove.

dopplershift commented on recipes/dfba/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+  run:
+    - cobra
+    - depinfo
+    - {{ compiler('cxx') }}
```
I don't think this needs a c++ compiler at *runtime*?

davidtourigny replied:
Hi @dopplershift the software uses a c++ compiler to compile a shared library at runtime (that is then loaded by the main program). I discussed with @Midnighter today whether it would be important to get a runtime test implemented to ensure this works cross-platform with conda-forge (as we have done [here](https://gitlab.com/davidtourigny/dynamic-fba)).

dopplershift replied:
AH, had no idea. Yeah, it would be really good to have a test that makes sure that this compilation part functions correctly.

Midnighter replied:
Hi @dopplershift, maybe you know how to do this best. I've included a `run_test.py` file in the recipe's directory which gets executed automatically so that's working great. However, I have a problem with getting the model file (`iJR904.xml.gz`) to be mounted in the directory where the tests are run. I have tried to do this by listing it in the `source_files` but that leads to the following error. What's the best way to achieve this?

```
Traceback (most recent call last):
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 175, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 100, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 156, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.8/site-packages/conda_build/api.py", line 186, in build
    return build_tree(
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 3068, in build_tree
    packages_from_this = build(metadata, stats,
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 2351, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 1657, in bundle_conda
    output['checksums'] = create_info_files(metadata, replacements, files, prefix=metadata.config.host_prefix)
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 1266, in create_info_files
    copy_test_source_files(m, join(m.config.info_dir, 'test'))
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 830, in copy_test_source_files
    raise RuntimeError(msg.format(pattern))
RuntimeError: Did not find any source_files for test with pattern iJR904.xml.gz
```

dopplershift replied:
Use [`files:`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#test-files) rather than `source_files:`. (I had to look this up.)

Midnighter replied:
Sweet, that did it thanks!

Midnighter commented:
Alright, with your help @dopplershift I've added a kind of end-to-end test which includes dynamic compilation. I've also addressed your other comment. Good to go from my point of view.

dopplershift commented on recipes/dfba/run_test.py:
```diff
@@ -0,0 +1,101 @@
+    0.0, 25.0, 0.1, ["EX_glc(e)", "EX_xyl_D(e)", "EX_etoh(e)"]
+)
+
+# TODO: Assert some value ranges for the results.
```
Did you want to do anything about this "TODO"?

Midnighter replied:
Not at the moment. I'm happy with knowing that the simulation completes successfully.
